### PR TITLE
Bail out of solve quicker on exception

### LINKED ISSUE #23008: `_fail_next_linear_convergence_check` does nothing anymore
## Reason
In c1f926412e48c448f718da8d760b9b998214f16f Fande took out `FEProblemBase::checkLinearConvergence` as it was not used in PETSc 3.10 and above. This mechanism was previously responsible for early termination of time steps in case exceptions were thrown during residual evaluation (`_fail_next_linear_convergence_check` is set to `true` in `checkExceptionAndStopSolve`).

## Design
Discuss! This is the reason why exceptions do not cut the time step as swiftly as they used to. 

## Impact
Fixing this would speed up difficult models where exceptions get thrown.

@oanaoana @bwspenc

### PR DESCRIPTION
**Pull request by dschwen**
This works for me and tremendously speeds up a production run.

Closes #23008

### CODE REVIEW

**Comment by friedmud** [comment]
I like it.  Can you please add a test so we can make sure not to lose this capability again?

File: framework/include/problems/FEProblemBase.h [hunk 1]
```diff
@@ -1998,6 +1998,9 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   unsigned int nlSysNum(const NonlinearSystemName & nl_sys_name) const;
 
+  /// Skip further residual evaluations and fail the next nonlinear convergence check
+  bool failNextNonlinearConvergenceCheck() { return _fail_next_nonlinear_convergence_check; }
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:2002]:
```suggestion
  /**
   * Skip further residual evaluations and fail the next nonlinear convergence check
   */
  bool failNextNonlinearConvergenceCheck() const { return _fail_next_nonlinear_convergence_check; }
```

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -7262,6 +7263,12 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
   TIME_SECTION("checkNonlinearConvergence", 5, "Checking Nonlinear Convergence");
   mooseAssert(_current_nl_sys, "This should be non-null");
 
+  if (_fail_next_nonlinear_convergence_check)
+  {
+    _fail_next_nonlinear_convergence_check = false;
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:7268]: Are both resets necessary? This is not all that important of a question

**Review by lindsayad — CHANGES_REQUESTED** [review]
I am curious what the screen output of a solve with a failed exception looks like now?

**Comment by dschwen** [comment]
> I am curious what the screen output of a solve with a failed exception looks like now?

```
+-progress-------------------------------------------------------------------------------------------+
|###########################################################.........................................|
+----------------------------------------------------------------------------------------------------+

Time Step 18, time = 363400, dt = 1800
 0 Nonlinear |R| = 1.513582e+04
   Solution, reference convergence variable norms:
   disp_x (grouped) : 352.952  disp_x (grouped) _ref: 353.007
   T:                 15131.7  T_ref:                 21431.1
      0 Linear |R| = 1.513582e+04
      1 Linear |R| = 4.871769e-11
checkExceptionAndStopSolve: Exceeded maximum iterations in material return mapping iterations.
At element 25604 _qp=1 Coordinates (x,y,z)=(-1.17505,  8.07672, 0.708922) block=1
 dt 2700 old cell disl: 1e+13 old wall disl: 5e+11 old effective strain: 0
 temp: 788.498 environmental: 0 trial stress into rom (MPa): 84.646 cell: 1e+13 wall: 5e+11 old strain: 0
  partition 2 weight: 0
  weights by tile, partition 1:  (0, 0)  (1, 0)  (2, 0)  (3, 0)  (4, 0)  (5, 0)  (6, 0)  (7, 0)  (8, 0)  (9, 0)  (10, 0)  (11, 0)  (12, 0)  (13, 0)  (14, 0)  (15, 0)  (16, 0)  (17, 0)  (18, 0)  (19, 0)  (20, 0)  (21, 0)  (22, 0)  (23, 0)  (24, 0)  (25, 0)  (26, 0)  (27, 0)  (28, 0)  (29, 0)  (30, 0)  (31, 0)  (32, 0)  (33, 0)  (34, 0) 

  nonstress weights by tile, partition 1:  (0, 0)  (1, 0)  (2, 0)  (3, 0)  (4, 0)  (5, 0)  (6, 0)  (7, 0)  (8, 0)  (9, 0)  (10, 0)  (11, 0)  (12, 0)  (13, 0)  (14, 0)  (15, 1)  (16, 0)  (17, 0)  (18, 0)  (19, 0)  (20, 0)  (21, 1)  (22, 0)  (23, 0)  (24, 0)  (25, 0)  (26, 0)  (27, 1)  (28, 0)  (29, 0)  (30, 0)  (31, 0)  (32, 0)  (33, 1)  (34, 0) 

In 1000 iterations the residual went from 4.12622e-05 to -1.08766e-05 in 'rom_stress_prediction'.

Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!
  Finished Solving                                                                       [ 11.08 s] [  -23 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.

Time Step 18, time = 362950, dt = 1350
 0 Nonlinear |R| = 1.134814e+04
   Solution, reference convergence variable norms:
   disp_x (grouped) : 232.987  disp_x (grouped) _ref: 233.049
   T:                 11345.7  T_ref:                 16066.7
      0 Linear |R| = 1.134814e+04
      1 Linear |R| = 3.465320e-11
 1 Nonlinear |R| = 9.197879e+00
   Solution, reference convergence variable norms:
   disp_x (grouped) : 0.501774  disp_x (grouped) _ref: 0.850385
   T:                 9.18418  T_ref:                 304.055
      0 Linear |R| = 9.197879e+00
      1 Linear |R| = 1.198220e-13
 2 Nonlinear |R| = 1.328746e-03
   Solution, reference convergence variable norms:
   disp_x (grouped) : 0.00130805  disp_x (grouped) _ref: 0.67784
   T:                 0.000233608  T_ref:                 302.547
      0 Linear |R| = 1.328746e-03
      1 Linear |R| = 2.983062e-16
...
```

File: framework/src/problems/FEProblemBase.C (hunk 2 above)
Comment (dschwen) [reply framework/src/problems/FEProblemBase.C:7268]: probably not. let me double check...

Comment (dschwen) [reply framework/src/problems/FEProblemBase.C:7268]: Hm, not clear to me. Two scenarios come to mind

* anyone overrides the `checkNonlinearConvergence` function - we'd need the reset in `solve` for MOOSE to still work
* anything relied on evaluating the residual between the last  `checkNonlinearConvergence` and the next call to `solve` (rescaling?)

Having two resets would be a bit more robust.

**Comment by dschwen** [comment]
Man, I thought I was taking crazy pills, because the tests for exception throwing that we have look like they are doing the right thing! But they all throw during the first residual evaluation of the first timestep. If you throw during a linear iteration you get a totally different picture:

## Pre-PR

```
Time Step 0, time = 0

Time Step 1, time = 0.01, dt = 0.01
MooseException thrown during residual calculation
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 1.732051e+00
      1 Linear |R| = 6.547838e-02
      2 Linear |R| = 3.330733e-16
 1 Nonlinear |R| = 2.238545e-08
      0 Linear |R| = 2.238545e-08
      1 Linear |R| = 8.462585e-10
      2 Linear |R| = 3.554904e-18
 2 Nonlinear |R| = 7.693103e-16
 Solve Converged!

Time Step 2, time = 0.02, dt = 0.01
 0 Nonlinear |R| = 1.723301e+00
      0 Linear |R| = 1.723301e+00
      1 Linear |R| = 7.502770e-10
 1 Nonlinear |R| = 1.227396e-09
 Solve Converged!
```


## With this PR

```
Time Step 0, time = 0

Time Step 1, time = 0.01, dt = 0.01
MooseException thrown during residual calculation
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 0.005, dt = 0.005
 0 Nonlinear |R| = 1.732051e+00
      0 Linear |R| = 1.732051e+00
      1 Linear |R| = 2.212251e-01
      2 Linear |R| = 6.665608e-16
 1 Nonlinear |R| = 8.187440e-09
 Solve Converged!
```

Without this PR the exception is thrown but the time step is never cut!

**Comment by dschwen** [comment]
Test added. That test would fail on vanilla MOOSE with the following CSV output

```
time,linear_steps
0,0
0.01,10
0.02,15
0.03,20
0.04,25
0.05,30
```

File: test/tests/misc/exception/tests [hunk 3]
```diff
@@ -89,16 +83,26 @@
...
+  []
+
+  [exception_transient]
+    type = RunApp
+    # type = 'CSVDiff'
+    input = 'exception_transient.i'
+
+    allow_warnings = true
+    allow_test_objects = true
+    requirement = "The system shall interrupt a solve immediately as soon as an exception is thrown."
```
Comment (GiudGiud) [inline test/tests/misc/exception/tests:106]: CSVDiff or runexception would do better

File: framework/include/problems/FEProblemBase.h [hunk 4]
```diff
@@ -1998,6 +1998,11 @@ class FEProblemBase : public SubProblem, public Restartable
    */
   unsigned int nlSysNum(const NonlinearSystemName & nl_sys_name) const;
 
+  /**
+   * Skip further residual evaluations and fail the next nonlinear convergence check
+   */
+  bool failNextNonlinearConvergenceCheck() { return _fail_next_nonlinear_convergence_check; }
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:2004]: Looks like you missed my suggested const addition

File: framework/src/problems/FEProblemBase.C (hunk 2 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:7268]: Yea this is fine

File: test/tests/misc/exception/tests (hunk 3 above)
Comment (dschwen) [reply test/tests/misc/exception/tests:106]: D'oh, of course it should be CSVDiff, that's why I checked the gold file in :-D

File: framework/include/problems/FEProblemBase.h (hunk 4 above)
Comment (dschwen) [reply framework/include/problems/FEProblemBase.h:2004]: Yeah, I did. Sorry!

**Review by lindsayad — DISMISSED** [review]

**Comment by dschwen** [comment]
Ok diffing the number of linear iterations is not a good strategy.

**Review by lindsayad — APPROVED** [review]

**Pull request merged**